golem/payload/defaults.py

```diff
@@ -56,12 +65,45 @@ class ActivityInfo(Payload):
     """
 
     expiration: Optional[datetime] = prop("golem.srv.comp.expiration", default=None)
+    """The datetime until which any started activities will last."""
+
+    lifetime: Optional[timedelta] = field(default=None)
+    """Lifetime of the activities from the moment the demand is placed.
+
+    Convenience property, used only in case the actual `expiration` is not provided.
+    """
+
     multi_activity: Optional[bool] = prop("golem.srv.caps.multi-activity", default=None)
     """Whether client supports multi_activity (executing more than one activity per agreement).
     """
 
+    def __post_init__(self):
+        if self.expiration and self.lifetime:
+            raise ValueError(
+                "Ambiguous definition - either the expiration or the lifetime must be provided."
+            )
+
+        # in case the expiration itself is not provided, set the default lifetime instead
+        # so that the expiration can be dynamically constructed when needed
+        if not self.expiration and not self.lifetime:
+            self.lifetime = DEFAULT_LIFETIME
+
+    async def build_properties_and_constraints(self) -> Tuple[Properties, Constraints]:
+        # we don't want to freeze the expiration, so we're making a copy here
+        if not self.expiration:
```

Comment:
`if not self.expiration and self.lifetime:`

Instead of `assert activity_info.lifetime` will allow to generate `ActivityInfo` with both `expiration` and `lifetime` set to None.

Reply:
not really, because `activity_info.lifetime` cannot be `None` anyway